File: 967-numbers-with-same-consecutive-differences/967-numbers-with-same-consecutive-differences.py
class Solution:
    def numsSameConsecDiff(self, n: int, k: int) -> List[int]:
        res = []
        if k == 0:
            return [int(str(d) * n) for d in range(1, 10)]
        def f(generated: int, last: int, remaining_digit_count: int):
            if remaining_digit_count == 0:
                res.append(generated)
                return
            next_num = last + k
            if next_num < 10:
                f(generated * 10 + next_num, next_num, remaining_digit_count - 1)
            next_num = last - k
            if next_num > -1:
                f(generated * 10 + next_num, next_num, remaining_digit_count - 1)
        for i in range(1, 10):
            f(i, i, n - 1)
        return res

File: 967-numbers-with-same-consecutive-differences/967-numbers-with-same-consecutive-differences.rs
impl Solution {
    pub fn nums_same_consec_diff(n: i32, k: i32) -> Vec<i32> {
        fn f(gen: i32, last: i32, left: i32, res: &mut Vec<i32>, k: &i32) {
            if left == 0 {
                return res.push(gen);
            }
            if *k == 0 {
                return f(gen * 10 + last, last, left - 1, res, k);
            }
            let next = last + k;
            if next < 10 {
                f(gen * 10 + next, next, left - 1, res, k);
            }
            let next = last - k;
            if next > -1 {
                f(gen * 10 + next, next, left - 1, res, k);
            }
        }
        let mut res: Vec<i32> = vec![];
        for i in 1..10 {
            f(i, i, n - 1, &mut res, &k);
        }
        res
    }
}
